### PR TITLE
chore: Reorganise API code and tests

### PR DESCRIPTION
**Pull request by tofumatt**
This just separates out a lot of the code in core/api. I was finding it a real pain to read that file and while we've moved to using separate files for new code I wanted to separate it out because the tests and API code is hard to make sense of right now.

This didn't take too long and is solely reorganising. If you think it's not worth the time though just say so and you can close this, it's not the hugest deal.

### CODE REVIEW

File: src/core/api/index.js [hunk 1]
```diff
@@ -162,111 +161,22 @@ export function callApi({
...
+export const login = authenticationApi.login;
+export const logOutFromServer = authenticationApi.logOutFromServer;
+export const startLoginUrl = authenticationApi.startLoginUrl;
+// Category APIs
+export const categories = categoriesApi.categories;
+// Featured Add-on APIs
+export const featured = featuredApi.featured;
+// Search and Autocomplete APIs
+export const autocomplete = searchApi.autocomplete;
+export const search = searchApi.search;
```
Comment (willdurand) [inline src/core/api/index.js]: Can we stop doing this maybe? See: https://github.com/mozilla/addons-frontend/pull/3027#discussion_r137825532.

Comment (tofumatt) [reply src/core/api/index.js]: Yes I would like that... but it was more work especially with normalizr. I'll see about removing it in another commit though!

Comment (willdurand) [reply src/core/api/index.js]: Can't we just add per API `import` where we rely on `core/api`?

Comment (tofumatt) [reply src/core/api/index.js]: We can but there are some bugs with normalizr and stuff in the tests. But I'll give it a go again... 😄

Comment (tofumatt) [reply src/core/api/index.js]: Okay, I just spent a few hours trying to remove these imports and going through the code/tests changing calls from `import * as api from 'core/api'` to more specific imports but it's a total mess. Even after some basic cleanup there are like 50 failing tests and the causes aren't obvious. I feel like this branch is a definite improvement and moving to importing the functions from each file will be a lot easier once we ditch normalizr (#2917).

So I'm just gonna propose the PR as it is. I think it's a better jumping off point to fix #2917 but I don't want to go down the rabbit hole any further for now.

Comment (willdurand) [reply src/core/api/index.js]: Oh ok, sorry for the extra work 😔

Comment (tofumatt) [reply src/core/api/index.js]: Oh, no worries. It was worth trying but I realised eventually it just wasn't worth pursuing right now 😄

File: src/core/api/index.js [hunk 2]
```diff
@@ -162,111 +161,22 @@ export function callApi({
...
-
-export const search = searchApi;
+// Schemas
+// TODO: Remove this when
+// https://github.com/mozilla/addons-frontend/issues/2917 is fixed.
+export const addon = addonApi.addonSchema;
+export const category = categoriesApi.categorySchema;
+
+// Add-on APIs
+export const fetchAddon = addonApi.fetchAddon;
```
Comment (kumar303) [inline src/core/api/index.js]: I don't see the benefit in aliasing all of these submodules. It just creates confusion and requires a bunch of boilerplate code. Why can't we just update all of the import statements? If the test failures are confusing, you could could push a commit with the aliases removed and I'll take a look.

File: src/core/api/authentication.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import url from 'url';
```
Comment (kumar303) [inline src/core/api/authentication.js]: All of these new files need `/* @flow */` at the top

**Comment by tofumatt** [comment]
My thought was it's easier to read the files and deal with the code/tests when it's not a tonne of barely-related functions inside the same file. It's easier at a glance to see "oh this is the file for search, this one for categories". Aside from using the `callApi` method there is little in common with these functions.

Aliasing things means the code works as before but is at least more organised... especially the tests. The aliases are there because of the way the tests do:

```js
import * as api from 'core/api';

const mockApi = sinon.mock(api);
```

**Comment by kumar303** [comment]
```js
import * as api from 'core/api';

const mockApi = sinon.mock(api);
```

This seems easy enough to change, no?

```js
// In test_search.js, for example...

import * as api from 'core/api/search';

const mockApi = sinon.mock(api);
```

If you need two mocks in the same test file then some tests probably need moving to their respective file. If it doesn't make sense to move them you can just create two mocks:

```js
import * as searchApi from 'core/api/search';
import * as categoriesApi from 'core/api/categories';

const mockSearchApi = sinon.mock(searchApi);
const mockCategoriesApi = sinon.mock(categoriesApi);
```

**Comment by tofumatt** [comment]
I didn't try that approach but normalizr would sometimes throw errors if I tried something similar. I'll try again but more than an hour's work and I'd say best to save this for later...

**Comment by tofumatt** [comment]
Bit rot. I'll leave the branch here as a template, maybe.

**Pull request closed without merging**